ntoskrnl/ex/locale.c

```diff
@@ -24,8 +24,250 @@ LANGID PsInstallUILanguageId = LANGIDFROMLCID(0x00000409);
 LCID PsDefaultThreadLocaleId = 0x00000409;
 LANGID PsDefaultUILanguageId = LANGIDFROMLCID(0x00000409);
 
+/* DEFINES *******************************************************************/
+
+#define BOGUS_LOCALE_ID 0xFFFF0000
+#define LCID_HEX_FORMAT_DIGIT 8
+
 /* PRIVATE FUNCTIONS *********************************************************/
 
+/**
+ * @brief
+ * Validates the registry data of a NLS locale.
+ *
+ * @param[in] LocaleData
+ * A pointer to partial information that contains
+ * the NLS locale data.
+ *
+ * @return
+ * Returns TRUE if the following conditions are met,
+ * otherwise FALSE is returned.
+ */
+static
+BOOLEAN
+ExpValidateNlsLocaleData(
+    _In_ PKEY_VALUE_PARTIAL_INFORMATION LocaleData)
+{
+    PWCHAR Data;
+
+    /* Is this a null-terminated string type? */
+    if (LocaleData->Type != REG_SZ)
+    {
+        return FALSE;
+    }
+
+    /* Does it have a consistent length? */
+    if (LocaleData->DataLength < 2)
+    {
+        return FALSE;
+    }
+
+    /* Is the locale set and null-terminated? */
+    Data = (PWSTR)LocaleData->Data;
+    if (Data[0] != L'1' || Data[1] != UNICODE_NULL)
+    {
+        return FALSE;
+    }
+
+    /* All of the conditions above are met */
+    return TRUE;
+}
+
+/**
+ * @brief
+ * Validates a NLS locale. Whether a locale is valid
+ * or not depends on the following conditions:
+ *
+ * - The locale must exist in the Locale key, otherwise
+ *   in the Alternate Sorts key;
+ *
+ * - The locale must exist in the Language Groups key, and
+ *   the queried value must be readable;
+ *
+ * - The locale registry data must be of REG_SIZE type,
+ *   has a consistent length and the locale is set.
+ *
+ * @param[in] LocaleId
+ * A locale identifier that corresponds to a specific
+ * locale to be validated.
+ *
+ * @return
+ * Returns STATUS_SUCCESS if the function has successfully
+ * validated the locale and it is valid. STATUS_OBJECT_NAME_NOT_FOUND
+ * is returned if the following locale does not exist on the system.
+ * A failure NTSTATUS code is returned otherwise.
+ */
+static
+NTSTATUS
+ExpValidateNlsLocaleId(
+    _In_ LCID LocaleId)
+{
+    NTSTATUS Status;
+    HANDLE NlsLocaleKey = NULL, AltSortKey = NULL, LangGroupKey = NULL;
+    OBJECT_ATTRIBUTES NlsLocalKeyAttrs, AltSortKeyAttrs, LangGroupKeyAttrs;
+    PKEY_VALUE_PARTIAL_INFORMATION BufferKey;
+    WCHAR ValueBuffer[20], LocalBuffer[20], ValueIdBuffer[20];
+    USHORT BufferLength;
+    PWCHAR Id;
+    ULONG i, IdLength, ReturnedLength;
+    UNICODE_STRING LocaleName;
+    UNICODE_STRING LocaleIdString;
+    static UNICODE_STRING NlsLocaleKeyPath = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\SYSTEM\\CurrentControlSet\\Control\\Nls\\Locale");
+    static UNICODE_STRING AltSortKeyPath = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\SYSTEM\\CurrentControlSet\\Control\\Nls\\Locale\\Alternate Sorts");
+    static UNICODE_STRING LangGroupPath = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\SYSTEM\\CurrentControlSet\\Control\\Nls\\Language Groups");
```

Comment:
You could instead open the NlsLocale Key, then use the obtained key as the parent one from which you open the two others.

Reply:
That would actually require that you open `\\Nls` which is what you mean (that would be your actual parent), then opening `Locale`, `Alternate Sorts` and then `Language Groups`, equaling to a total of 4 opened keys. I don't really see the benefit of this.